ts_src/payments/taprootutils.ts

```diff
@@ -0,0 +1,144 @@
+import { Buffer as NBuffer } from 'buffer';
+import * as bcrypto from '../crypto';
+
+import { varuint } from '../bufferutils';
+import { TaprootLeaf } from '../types';
+
+const TAP_LEAF_TAG = 'TapLeaf';
+const TAP_BRANCH_TAG = 'TapBranch';
+const TAP_TWEAK_TAG = 'TapTweak';
+
+export const LEAF_VERSION_TAPSCRIPT = 0xc0;
+
+export function rootHashFromPath(
+  controlBlock: Buffer,
+  tapLeafMsg: Buffer,
+): Buffer {
+  const k = [tapLeafMsg];
+  const e = [];
+
+  const m = (controlBlock.length - 33) / 32;
+
+  for (let j = 0; j < m; j++) {
+    e[j] = controlBlock.slice(33 + 32 * j, 65 + 32 * j);
+    if (k[j].compare(e[j]) < 0) {
+      k[j + 1] = tapBranchHash(k[j], e[j]);
+    } else {
+      k[j + 1] = tapBranchHash(e[j], k[j]);
+    }
+  }
+
+  return k[m];
+}
+
+export interface HashTree {
+  hash: Buffer;
+  left?: HashTree;
+  right?: HashTree;
+}
+
+/**
+ * Build the hash tree from the scripts binary tree.
+ * The binary tree can be balanced or not.
+ * @param scriptTree - is a list representing a binary tree where an element can be:
+ *  - a taproot leaf [(output, version)], or
+ *  - a pair of two taproot leafs [(output, version), (output, version)], or
+ *  - one taproot leaf and a list of elements
+ */
+export function toHashTree(scriptTree: TaprootLeaf[]): HashTree {
```

Comment:
it seems that `scriptTree` type should be `Array<[TaprootLeaf, TaprootLeaf] | TaprootLeaf>` ?

Reply:
Solved, thanks!